quant: add q_batchnorm_1d op

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42546 quant: add quantized batchnorm 1d module
* **#42491 quant: add q_batchnorm_1d op**

Summary:

Hooks up quantized batchnorm_1d to the quantized_bn kernel. Eager mode
hookup will be in a future PR, and graph mode should work after this PR.

Note: currently the implementation is ~2x slower on the benchmark than q_batch_norm2d
because we convert back to contiguous memory format at the end, since
channels_last is only defined for rank >= 4. If further optimization is
needed, that can be a separate PR (will need the NHWC folks to see if
there is a workaround).  Meanwhile, having this is better than not having anything.

Context: There have been both internal and external requests for various
quantized BN1d use cases.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_batch_norm_1d_2d_3d
python test/test_quantization.py TestQuantizedOps.test_batch_norm_1d_2d_3d_relu
python test/test_quantization.py TestQuantizeJitOps.test_qbatch_norm

// performance:
// https://gist.github.com/vkuzo/73a07c0f24c05f5804990d9ebfaecf5e

```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22926254](https://our.internmc.facebook.com/intern/diff/D22926254)

vkuzo committed:
quant: add q_batchnorm_1d kernel

Summary:

Hooks up quantized batchnorm_1d to the quantized_bn kernel. Eager mode
hookup will be in a future PR, and graph mode should work after this PR.

Note: currently the implementation is ~2x slower on the benchmark than q_batch_norm2d
because we convert back to contiguous memory format at the end, since
channels_last is only defined for rank >= 4. If further optimization is
needed, that can be a separate PR (will need the NHWC folks to see if
there is a workaround).

Context: There have been both internal and external requests for various
quantized BN1d use cases.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_batch_norm_1d_2d_3d
python test/test_quantization.py TestQuantizedOps.test_batch_norm_1d_2d_3d_relu
python test/test_quantization.py TestQuantizeJitOps.test_qbatch_norm

// performance:
// https://gist.github.com/vkuzo/73a07c0f24c05f5804990d9ebfaecf5e

```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
fix nit on "quant: add q_batchnorm_1d kernel"


Summary:

Hooks up quantized batchnorm_1d to the quantized_bn kernel. Eager mode
hookup will be in a future PR, and graph mode should work after this PR.

Note: currently the implementation is ~2x slower on the benchmark than q_batch_norm2d
because we convert back to contiguous memory format at the end, since
channels_last is only defined for rank >= 4. If further optimization is
needed, that can be a separate PR (will need the NHWC folks to see if
there is a workaround).  Meanwhile, having this is better than not having anything.

Context: There have been both internal and external requests for various
quantized BN1d use cases.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_batch_norm_1d_2d_3d
python test/test_quantization.py TestQuantizedOps.test_batch_norm_1d_2d_3d_relu
python test/test_quantization.py TestQuantizeJitOps.test_qbatch_norm

// performance:
// https://gist.github.com/vkuzo/73a07c0f24c05f5804990d9ebfaecf5e

```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp:
```diff
@@ -39,6 +39,105 @@ void compute_fused_params(
+  // Remove the fake dimension, and go back to contiguous format
+  // (since there is no 4th channel). Note, this has a performance
+  // cost.
+  return qy.contiguous(MemoryFormat::Contiguous).squeeze(-1);
```
checked with Vitaly, we can add `MemoryFormat::ChannelsLast1d` as a future PR if we need to optimize this further.

jerryzh168 replied:
does squeeze affect the `ChannelsLast1d`? i.e. can squeeze preserve the channel order

vkuzo replied:
sorry, not 100% sure of the question, are you asking if it would work correctly if we squeeze without converting back to contiguous, and would it remove the correct dim?

jerryzh168 replied:
I mean does this implementation affect the memory layout of the Tensor.
let's say we implement the qbatch_norm_1d natively without squeeze, I assume that can preserve the memory layout of the Tensor, that is given a Tensor with non-contiguous NHWC layout the output would still be NHWC. and does using squeeze here mean that we need to make the Tensor contiguous?

vkuzo replied:
ah, sure.  Here is my current understanding:
* the cpu intrinsics used to make the kernel fast require the input data for a channel to be contiguous (because they do SIMD reads on elements which are adjacent in memory).  I'd guess that trying to write a kernel for BN without requiring channels to be contiguous would be slower than what we currently have.  Didn't try it though.
* the current approach is a hack to reuse the 2d kernel (add a dim, do kernel in NHWC, remove the dim).  The only reason we convert back from NHWC to contiguous at the end is because "3 channel channels-last" is not defined yet.  The data would be in the right format (and tests on this diff actually pass with 3-channel NHWC), but the scaffolding just isn't there for this data format to be read correctly by whoever will use this tensor.  
* I got an estimate of "1-2 eng weeks, high risk" to make the new memory format a first class citized from vfedunin, so we can choose to take that path in the future if the customer team really asks for it (and if this becomes a perf bottleneck)

sorry, still not 100% I answered your question - lmk if the ^ doesn't cover it

jerryzh168 replied:
Thanks for the clarification. I think that answers my question. My question is mostly that let's say all the previous ops are working with NHWC order and all the ops after this batchnorm is working with NHWC order, does this batchnorm break this? I guess the answer is yes, in this case. So basically here we need to go from NHWC to NCHW and then in the next op we'll go from NCHW to NHWC.

vkuzo replied:
ah, I see.  Sorry, now I fully understand the question.  I believe the way it's supposed to work is
* each op has to be able to work with either NHWC or NCHW input, so as long as the tensor is properly tagged with memory format then it's guaranteed a correct result.  If it doesn't support some input, it should throw an error.
* currently there are manual optimizations to make defaults be sane, and I think those manual guidelines ask to use NHWC if possible (in this case it's not possible yet because we need ChannelsLast1d first).
* in the future there might be graph level optimizations (i.e. converting nhwc_kernel -> nhwc_to_nchw -> ncwh_to_nhwc -> nhwc_kernel to just be nhwc_kernel -> nhwc_kernel).  I don't think that exists yet.

cc @kimishpatel  since he might know a more updated version of the guidelines on this.

vkuzo replied:
btw, for batchnorm1d specifically, there shouldn't be nhwc before and after, because of the same issue that `MemoryFormat::ChannelsLast1d` does not exist yet.

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -1969,9 +1942,9 @@ def test_instance_norm(self):
 
     @skipIfNoFBGEMM
-    def test_batch_norm2d_relu(self):
+    def test_batch_norm_1d_2d_3d_relu(self):
```
nit: maybe just test_batch_norm_relu and add dimensions in the docstring if needed.

vkuzo replied:
sure, sounds good

jerryzh168 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2017,17 +1993,17 @@ def test_batch_norm2d_relu(self):
 
     @skipIfNoFBGEMM
-    def test_batch_norm3d(self):
+    def test_batch_norm_1d_2d_3d(self):
```
same here

jerryzh168 reviewed (approved):
Looks good

vkuzo committed:
update test names on "quant: add q_batchnorm_1d op"


Summary:

Hooks up quantized batchnorm_1d to the quantized_bn kernel. Eager mode
hookup will be in a future PR, and graph mode should work after this PR.

Note: currently the implementation is ~2x slower on the benchmark than q_batch_norm2d
because we convert back to contiguous memory format at the end, since
channels_last is only defined for rank >= 4. If further optimization is
needed, that can be a separate PR (will need the NHWC folks to see if
there is a workaround).  Meanwhile, having this is better than not having anything.

Context: There have been both internal and external requests for various
quantized BN1d use cases.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_batch_norm_1d_2d_3d
python test/test_quantization.py TestQuantizedOps.test_batch_norm_1d_2d_3d_relu
python test/test_quantization.py TestQuantizeJitOps.test_qbatch_norm

// performance:
// https://gist.github.com/vkuzo/73a07c0f24c05f5804990d9ebfaecf5e

```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22926254](https://our.internmc.facebook.com/intern/diff/D22926254)

[ghstack-poisoned]
